rules/integrate-identityserver-with-an-existing-user-store/rule.md

```diff
@@ -0,0 +1,129 @@
+---
+type: rule
+title: Do you know how to migrate an existing user store to IdentityServer?
+uri: migrate-an-existing-user-store-to-identityserver
+authors:
+  - title: "Dhruv Mathur "
+    url: https://www.ssw.com.au/people/dhruv/
+created: 2023-10-31T04:31:12.396Z
+guid: 38a5988b-1740-4120-840d-116ad6e91566
+---
+When integrating an Identity Server with an existing ASP .NET Core application, challenges arise due to different user identification systems. 
+
+On the Identity Server side, users are typically recognised by a unique SubId within their issued token after authentication. In contrast, an application's existing user store might use its own unique user ID, possibly combined with other data. 
+
+The above discrepancy creates the need to effectively map or correlate the the user with SubId from the Identity Server to the corresponding user with user ID within the app's user store.
+
+## Two essential scenarios arise when integrating the Identity Server:
+
+### 1. Pre-existing Company Users
+
+#### **Technical Integration & Benefits**
+
+**Addressing users already registered with company emails in the existing application user store now authenticating through the Identity Server:**
+
+* **SubId Check**:
+
+  * Begin by verifying if the user has an external login associated with the SubId from the Identity Server in your application's user store. If found, proceed with authentication by handling the request gracefully.
+
+```csharp
+var existingUserByExternalLogin = await _userManager.FindByLoginAsync("IdentityServer", subId);
+```
+
+::: good
+Figure: Retrieving existing user by using the associated external login provider info which in this case is IdentityServer.
+:::
+
+* **Existing Users by Email Verification**:
+
+  * If there's no associated SubId, check if the email (provided by the Identity Server during authentication available as one of the claims in the JWT token) exists in your application's user store.
+
+```csharp
+var userByUserName = await _userManager.FindByEmailAsync(emailFromIdentityServer);
+```
+
+* For users known to your application but not authenticated via the Identity Server:
+
+  * Retrieve the SubId from the Identity Server.
+  * Use ASP .NET Core Identity's [`AddLoginAsync()`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.usermanager-1.addloginasync?view=aspnetcore-8.0)  method to associate this SubId as an external login with the user's record.
```

Comment:
Gold. This is the core learning point of this rule. Can we emphasise this at all?